src/elfpy/pricing_models/base.py

```diff
@@ -164,14 +164,106 @@ def calc_share_reserves(
         .. todo:: Write a test for this function
         """
 
+        # y = (z / 2) * (mu * (1 + rt)**(1/tau) - c)
+        # z = (2 * y) / (mu * (1 + rt)**(1/tau) - c)
         # Only want to renormalize time for APR ("annual", so hard coded to 365)
         # Don't want to renormalize stretched time
         annualized_time = time_utils.norm_days(time_remaining.days, 365)
-        share_reserves = bond_reserves / (
-            init_share_price * (1 - target_apr * annualized_time) ** (1 / time_remaining.stretched_time)
-        )  # z = y / (mu * (1 - rt)**(1/tau))
+        share_reserves = (
+            2
+            * bond_reserves
+            / (
+                init_share_price * (1 - target_apr * annualized_time) ** (1 / time_remaining.stretched_time)
+                - init_share_price
+            )
+        )
+
         return share_reserves
 
+    def calc_base_for_target_apr(
+        self,
+        target_apr: float,
+        bond: float,
+        position_duration: StretchedTime,
+        share_price: float = 1.0,
+    ) -> float:
+        """Returns the base required to buy the given bonds at the target APR
+        For a long, this is maximum amount of base in required to get the given bonds out.
+        For a short, this is the minimum amount of base out for the given bonds in.
+
+        Parameters
+        ----------
+        target_apr : float
+            Target fixed APR in decimal units (for example, 5% APR would be 0.05)
+        bond_out: float
+            The amount of bonds to purchase
+        position_duration: StretchedTime
+            The term length of the bond
+        share_price : float
+            The current share price
+
+        Returns
+        -------
+        float
+            The base in to get bond_out bonds at the target_apr.
+
+        .. todo:: Write a test for this function
+        """
+
+        # delta_y / delta_z = p = 1 - r
+        # delta_z = delta_y / (1 - r)
+        # delta_x = c * delta_y / (1 - r)
+
+        # Only want to renormalize time for APR ("annual", so hard coded to 365)
+        # Don't want to renormalize stretched time
+        annualized_time = time_utils.norm_days(position_duration.days, 365)
+        base_in = share_price * bond / (1 - target_apr * annualized_time)
+
+        assert base_in >= 0, "base value negative"
```

Comment:
would this happen as a result of inappropriate input values? If so, shouldn't the assert be on the input ranges? If this could happen with reasonable values, isn't it better to put bounds on `base_in` such that it is always valid. Seems weird to put an assert on a computed value.

Reply:
This was by no means meant to be exhaustive.  Check the other functions in base.py.  If we have exhaustive asserts in our base level calc methods, we are adding a lot of steps to run the code.  This one assert does a minimum level of sanity checking for the output which is nice I think.  I actually could be convinced to take the assert _out_ and throw an error in the test instead, but this made testing a little easier.

Reply:
I see, ok I'm good with it.